Build failure: mesa

### Nixpkgs version

- Unstable (25.11)

### Steps to reproduce

On platforms without JIT support (see [here](https://github.com/llvm/llvm-project/blob/a4767e63eebfa1ae0065bdd8813df3839f62b461/llvm/CMakeLists.txt#L562)), `mesa` fails to build on staging-next due to an error related to LLVM.



### Can Hydra reproduce this build failure?

Hydra doesn’t try to build the package.

### Link to Hydra build job

_No response_

### Relevant log output

```console
[1166/5080] Compiling C++ object src/gallium/auxiliary/libgallium.a.p/gallivm_lp_bld_init_orc.cpp.o
FAILED: [code=1] src/gallium/auxiliary/libgallium.a.p/gallivm_lp_bld_init_orc.cpp.o
g++ -Isrc/gallium/auxiliary/libgallium.a.p -Isrc/gallium/auxiliary -I../src/gallium/auxiliary -Isrc/loader -I../src/loader -I../src/gallium/include -Isrc -I../src -Iinclude -I../include -I../src/gal
lium/auxiliary/util -Isrc/compiler/nir -I../src/compiler/nir -Isrc/compiler -I../src/compiler -Isrc/util/format -I/nix/store/pprs1czv105af25j11ryf5z3fzli22q6-libdrm-2.4.125-dev/include -I/nix/store/
pprs1czv105af25j11ryf5z3fzli22q6-libdrm-2.4.125-dev/include/libdrm -I/nix/store/cq8jyz8c02z44m68rc89g5575d8jf2fm-zlib-1.3.1-dev/include -I/nix/store/0zn8yhvc3fj4v6w2ixjndgqwvxv70wdf-zstd-1.5.7-dev/i
nclude -I/nix/store/rq216qq7wq5wnbvyc36vzhaf8gb47dqd-libunwind-1.8.3-dev/include -fvisibility=hidden -fdiagnostics-color=always -DNDEBUG -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c++17 '-DPACK
AGE_VERSION="25.2.3"' '-DPACKAGE_BUGREPORT="https://gitlab.freedesktop.org/mesa/mesa/-/issues"' -DHAVE_OPENGL=1 -DHAVE_OPENGL_ES_1=1 -DHAVE_OPENGL_ES_2=1 -DHAVE_ASAHI -DHAVE_CROCUS -DHAVE_D3D12 -DHA
VE_ETNAVIV -DHAVE_FREEDRENO -DHAVE_I915 -DHAVE_IRIS -DHAVE_LIMA -DHAVE_LLVMPIPE -DHAVE_NOUVEAU -DHAVE_PANFROST -DHAVE_R300 -DHAVE_R600 -DHAVE_RADEONSI -DHAVE_SOFTPIPE -DHAVE_SVGA -DHAVE_TEGRA -DHAVE
_V3D -DHAVE_VC4 -DHAVE_VIRGL -DHAVE_ZINK -DHAVE_SWRAST -DHAVE_NVK -DVIDEO_CODEC_VC1DEC=1 -DVIDEO_CODEC_H264DEC=1 -DVIDEO_CODEC_H264ENC=1 -DVIDEO_CODEC_H265DEC=1 -DVIDEO_CODEC_H265ENC=1 -DVIDEO_CODEC
_AV1DEC=1 -DVIDEO_CODEC_AV1ENC=1 -DVIDEO_CODEC_VP9DEC=1 -DHAVE_X11_PLATFORM -DHAVE_WAYLAND_PLATFORM -DHAVE_SURFACELESS_PLATFORM -DHAVE_DRM_PLATFORM -DHAVE_XCB_PLATFORM -DUSE_LIBGLVND=1 -DUSE_D3D12_P
REVIEW_HEADERS=0 -DHAVE_GALLIUM_D3D12_VIDEO -DHAVE_ST_VDPAU -DHAVE_VA_SURFACE_ATTRIB_DRM_FORMAT_MODIFIERS -DGLX_INDIRECT_RENDERING -DGLX_DIRECT_RENDERING -DGLX_USE_DRM -DGLAPI_EXPORT_PROTO_ENTRY_POI
NTS=0 -DALLOW_KCMP -DMESA_DEBUG=0 -DENABLE_SHADER_CACHE -DHAVE___BUILTIN_BSWAP32 -DHAVE___BUILTIN_BSWAP64 -DHAVE___BUILTIN_CLZ -DHAVE___BUILTIN_CLZLL -DHAVE___BUILTIN_CTZ -DHAVE___BUILTIN_EXPECT -DH
AVE___BUILTIN_FFS -DHAVE___BUILTIN_FFSLL -DHAVE___BUILTIN_POPCOUNT -DHAVE___BUILTIN_POPCOUNTLL -DHAVE___BUILTIN_UNREACHABLE -DHAVE___BUILTIN_TYPES_COMPATIBLE_P -DHAVE_FUNC_ATTRIBUTE_CONST -DHAVE_FUN
C_ATTRIBUTE_FLATTEN -DHAVE_FUNC_ATTRIBUTE_MALLOC -DHAVE_FUNC_ATTRIBUTE_PURE -DHAVE_FUNC_ATTRIBUTE_UNUSED -DHAVE_FUNC_ATTRIBUTE_WARN_UNUSED_RESULT -DHAVE_FUNC_ATTRIBUTE_WEAK -DHAVE_FUNC_ATTRIBUTE_FOR
MAT -DHAVE_FUNC_ATTRIBUTE_PACKED -DHAVE_FUNC_ATTRIBUTE_RETURNS_NONNULL -DHAVE_FUNC_ATTRIBUTE_ALIAS -DHAVE_FUNC_ATTRIBUTE_NORETURN -DHAVE_FUNC_ATTRIBUTE_OPTIMIZE -DHAVE_FUNC_ATTRIBUTE_VISIBILITY -DHA
VE_UINT128 -DHAVE_REALLOCARRAY -DHAVE_FMEMOPEN -DUSE_GCC_ATOMIC_BUILTINS -DUSE_LOONGARCH64_ASM -DMAJOR_IN_SYSMACROS -DHAS_SCHED_H -DHAS_SCHED_GETAFFINITY -DHAVE_LINUX_FUTEX_H -DHAVE_ENDIAN_H -DHAVE_
DLFCN_H -DHAVE_SYS_SHM_H -DHAVE_SYS_INOTIFY_H -DHAVE_LINUX_UDMABUF_H -DHAVE_STRTOF -DHAVE_MKOSTEMP -DHAVE_MEMFD_CREATE -DHAVE_RANDOM_R -DHAVE_FLOCK -DHAVE_STRTOK_R -DHAVE_GETRANDOM -DHAVE_POSIX_FALL
OCATE -DHAVE_SECURE_GETENV -DHAVE_GNU_QSORT_R -DHAVE_STRUCT_TIMESPEC -DHAVE_PROGRAM_INVOCATION_NAME -DHAVE_ISSIGNALING -DHAVE_POSIX_MEMALIGN -DHAVE_DIRENT_D_TYPE -DHAVE_STRTOD_L -DHAVE_DLADDR -DHAVE
_DL_ITERATE_PHDR -DHAVE_ZLIB -DHAVE_ZSTD -DHAVE_COMPRESSION -DHAVE_PTHREAD -DHAVE_PTHREAD_SETAFFINITY -DHAVE_LIBDRM -DHAVE_LIBUDEV '-DMESA_LLVM_VERSION_STRING="21.1.1"' -DLLVM_IS_SHARED=1 -DLLVM_AVA
ILABLE=1 -DDRAW_LLVM_AVAILABLE=1 -DAMD_LLVM_AVAILABLE=1 -DGALLIVM_USE_ORCJIT=1 -DHAVE_SPIRV_TOOLS -DUSE_LIBELF -DTHREAD_SANITIZER=0 -DHAVE_LIBUNWIND -DWL_HIDE_DEPRECATED -DHAVE_WL_DISPATCH_QUEUE_TIM
EOUT -DHAVE_WL_CREATE_QUEUE_WITH_NAME -DXCB_KEYSYMS_AVAILABLE -DHAVE_X11_DRM -DHAVE_DRI -DHAVE_DRI2 -DHAVE_DRI3_EXPLICIT_SYNC -DHAVE_DRISW_KMS -DHAVE_GALLIUM_EXTRA_HUD=1 -DHAVE_LIBSENSORS=1 -mtls-di
alect=desc -Werror=return-type -Werror=empty-body -Wmisleading-indentation -Wno-non-virtual-dtor -Wno-missing-field-initializers -Wno-format-truncation -fno-math-errno -fno-trapping-math -flifetime-
dse=1 -Wno-unknown-pragmas -Werror=format -Wformat-security -ffunction-sections -fdata-sections -fPIC -pthread -D_GNU_SOURCE -DEXPERIMENTAL_KEY_INSTRUCTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_
MACROS -D__STDC_LIMIT_MACROS -Werror=pointer-arith -Werror=vla -MD -MQ src/gallium/auxiliary/libgallium.a.p/gallivm_lp_bld_init_orc.cpp.o -MF src/gallium/auxiliary/libgallium.a.p/gallivm_lp_bld_init
_orc.cpp.o.d -o src/gallium/auxiliary/libgallium.a.p/gallivm_lp_bld_init_orc.cpp.o -c ../src/gallium/auxiliary/gallivm/lp_bld_init_orc.cpp
../src/gallium/auxiliary/gallivm/lp_bld_init_orc.cpp: In constructor '{anonymous}::LPJit::LPJit()':
../src/gallium/auxiliary/gallivm/lp_bld_init_orc.cpp:343:13: error: cannot convert '{anonymous}::LPJit::LPJit()::<lambda(llvm::orc::ExecutionSession&, const llvm::Triple&)>' to 'llvm::orc::LLJITBuil
derState::ObjectLinkingLayerCreator' {aka 'std::function<llvm::Expected<std::unique_ptr<llvm::orc::ObjectLayer> >(llvm::orc::ExecutionSession&)>'}
  343 |             [&](ExecutionSession &ES, const llvm::Triple &TT) {
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |             |
      |             {anonymous}::LPJit::LPJit()::<lambda(llvm::orc::ExecutionSession&, const llvm::Triple&)>
  344 |                return std::make_unique<ObjectLinkingLayer>(
      |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  345 |                   ES, ExitOnErr(llvm::jitlink::InProcessMemoryManager::Create()));
      |                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  346 |             })
      |             ~
In file included from ../src/gallium/auxiliary/gallivm/lp_bld_init_orc.cpp:38:
/nix/store/jfqvhs21wkqwfnlngwxyrm6s33ahb5dx-llvm-21.1.1-dev/include/llvm/ExecutionEngine/Orc/LLJIT.h:408:52: note:   initializing argument 1 of 'SetterImpl& llvm::orc::LLJITBuilderSetters< <template
-parameter-1-1>, <template-parameter-1-2>, <template-parameter-1-3> >::setObjectLinkingLayerCreator(llvm::orc::LLJITBuilderState::ObjectLinkingLayerCreator) [with JITType = llvm::orc::LLJIT; SetterI
mpl = llvm::orc::LLJITBuilder; State = llvm::orc::LLJITBuilderState; llvm::orc::LLJITBuilderState::ObjectLinkingLayerCreator = std::function<llvm::Expected<std::unique_ptr<llvm::orc::ObjectLayer> >(
llvm::orc::ExecutionSession&)>]'
  408 |       LLJITBuilderState::ObjectLinkingLayerCreator CreateObjectLinkingLayer) {
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
../src/gallium/auxiliary/gallivm/lp_bld_init_orc.cpp: In function 'bool init_gallivm_state(gallivm_state*, const char*, lp_context_ref*, lp_cached_code*)':
../src/gallium/auxiliary/gallivm/lp_bld_init_orc.cpp:555:23: error: 'LLVMOrcThreadSafeContextGetContext' was not declared in this scope; did you mean 'LLVMOrcThreadSafeContextRef'?
  555 |    gallivm->context = LLVMOrcThreadSafeContextGetContext(context->ref);
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                       LLVMOrcThreadSafeContextRef
```

### Additional context

Looks like https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/37027 could be the fix.

### System metadata

 - system: `"loongarch64-linux"`
 - host os: `Linux 6.16.9, NixOS, 25.11 (Xantusia), 25.11.20250927.6a6f257`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.31.2`
 - channels(root): `"nixos"`
 - nixpkgs: `not found`

### Notify maintainers

@K900 @vcunat

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/libraries/mesa/default.nix
+++ b/pkgs/development/libraries/mesa/default.nix
@@ -144,6 +144,8 @@ stdenv.mkDerivation {
 
   patches = [
     ./opencl.patch
+    # https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/37027
+    ./gallivm-llvm-21.patch
   ];
 
   postPatch = ''

--- a/pkgs/development/libraries/mesa/gallivm-llvm-21.patch
+++ b/pkgs/development/libraries/mesa/gallivm-llvm-21.patch
@@ -1,0 +1,53 @@
+From eca19331d94005485de5246cfa87a21621486cd8 Mon Sep 17 00:00:00 2001
+From: no92 <no92.mail@gmail.com>
+Date: Wed, 27 Aug 2025 16:02:31 +0200
+Subject: [PATCH] gallivm: support LLVM 21
+
+LLVM PR#146819 changed the signature of `setObjectLinkingLayerCreator`,
+dropping the Triple argument. The PR was first included in the LLVM 21
+series, and the new signature is gated behind a version check for that.
+
+`LLVMOrcThreadSafeContextGetContext` was removed in LLVM commit b18e5b6,
+and the ORC examples in the LLVM tree seem to just create a context
+instead, which we replicate here.
+
+With this commit, mesa successfully builds the llvmpipe gallium driver
+on riscv64 with LLVM 21.1.0.
+
+Closes: https://gitlab.freedesktop.org/mesa/mesa/-/issues/13785
+Closes: https://gitlab.freedesktop.org/mesa/mesa/-/issues/13852
+
+Reviewed-by: David Heidelberg <david@ixit.cz>
+---
+ src/gallium/auxiliary/gallivm/lp_bld_init_orc.cpp | 7 ++++++-
+ 1 file changed, 6 insertions(+), 1 deletion(-)
+
+diff --git a/src/gallium/auxiliary/gallivm/lp_bld_init_orc.cpp b/src/gallium/auxiliary/gallivm/lp_bld_init_orc.cpp
+index 3d2b8cf81bc0c..0be69b02b6eef 100644
+--- a/src/gallium/auxiliary/gallivm/lp_bld_init_orc.cpp
++++ b/src/gallium/auxiliary/gallivm/lp_bld_init_orc.cpp
+@@ -340,7 +340,12 @@ LPJit::LPJit() :jit_dylib_count(0) {
+          .setJITTargetMachineBuilder(std::move(JTMB))
+ #ifdef USE_JITLINK
+          .setObjectLinkingLayerCreator(
++#if LLVM_VERSION_MAJOR >= 21
++            /* LLVM 21 removed the Triple argument */
++            [&](ExecutionSession &ES) {
++#else
+             [&](ExecutionSession &ES, const llvm::Triple &TT) {
++#endif
+                return std::make_unique<ObjectLinkingLayer>(
+                   ES, ExitOnErr(llvm::jitlink::InProcessMemoryManager::Create()));
+             })
+@@ -552,7 +557,7 @@ init_gallivm_state(struct gallivm_state *gallivm, const char *name,
+    gallivm->cache = cache;
+ 
+    gallivm->_ts_context = context->ref;
+-   gallivm->context = LLVMOrcThreadSafeContextGetContext(context->ref);
++   gallivm->context = LLVMContextCreate();
+ 
+    gallivm->module_name = LPJit::get_unique_name(name);
+    gallivm->module = LLVMModuleCreateWithNameInContext(gallivm->module_name,
+-- 
+GitLab
+

